fendermint/vm/message/src/signed.rs

```diff
@@ -104,14 +110,22 @@ impl SignedMessage {
         // so at least for now they are easy to tell apart: any `f410` address is coming from Ethereum API and must have
         // been signed according to the Ethereum scheme, and it could not have been signed by an `f1` address, it doesn't
         // work with regular accounts.
+        //
+        // We detect the case where the recipient is not an ethereum address. If that is the case then use regular signing rules,
+        // which should allow messages from ethereum accounts to go to any other type of account, e.g. custom Wasm actors.
         match maybe_eth_address(&message.from) {
-            Some(addr) => {
+            Some(addr) if is_eth_addr_compat(&message.to) => {
```

Comment:
Great question :thinking: 

It is certainly possible to invoke an ethereum contract [using a masked ID](https://github.com/consensus-shipyard/ipc/blob/c3e4c020d52decfbc57eb6e2618c947f3f7b29a7/fendermint/vm/actor_interface/src/eam.rs#L82-L86), which is an ethereum account hash that looks like `0xff0000...0123`; such an address will go in the `message.to` as an `f0` address. There is some caveat but I forgot what it was.

For a moment I thought, oh no, that means you can't invoke your custom Wasm actor you deployed in genesis. However, that's not the use case we wanted to enable here; rather we simply want to transfer tokens to an `f1` address from the `f410` address (which we know we're dealing with because of `Some(addr)`), and then use _that_ `f1` address to interact with the custom wasm actor. 

We could try to invoke the custom wasm actor from `f410`, and it was always possible, except that the guard you pointed out demanded that we do so with the `InvokeContract` method number, which is unlikely what we want, so _effectively_ it doesn't allow it.

Reply:
I was going to look up examples of invoking actors through their f0 address in IPC, which is possible, but also made me find the [caveat](https://github.com/consensus-shipyard/ipc/blob/f9bda9e83dcd6c7175a25df604f16aec40ffa973/fendermint/vm/interpreter/src/fvm/state/ipc.rs#L63-L66) I forgot, that they don't work very well with internal "who called me" checks.

Reply:
> rather we simply want to transfer tokens to an f1 address from the f410 address (which we know we're dealing with because of Some(addr)), and then use that f1 address to interact with the custom wasm actor.

I think this will work... I realized looking at your new smoke test that the `f410` address and the `f1` address _can be derived from the same wallet_. So, we'd just hide that from the user. 

It could still be a little cumbersome though... imagine you're a user and have funded yourself in the subnet via some UI. Behind the scenes, this UI would fund using the `f410` address _and_ send those funds to the cooresponsing `f1` address. Now, you go to lookup your balance on a block explorer using your ETH address (the only one you know of), and it won't reflect your subnet balance. You'd see that it sent some funds elsewhere, but that might be confusing. Anyway, it's likely solvable with UIs... just explaining the UX challenge.

Reply:
Yeah, I don't know, maybe send all the funds to the f1 address and pretend the f410 doesn't exist, if it's not needed for anything else.

FWIW I was told it was best to avoid using the same key for both, but again unless they tell me why... :man_shrugging:

Reply:
Since this handling is client-side, I think we could also make an exception in a subnet specific SDK that performs a check here on `to` like `is_eth_addr_compat`, but excludes masked IDs. 

Thanks for the help as always.

Reply:
Yes and to be fair there is a case in the Ethereum API where it [looks at the actor type](https://github.com/consensus-shipyard/ipc/blob/adcbc084972d662dddfdb70229928233dac67f42/fendermint/eth/api/src/apis/eth.rs#L757). Signature check is stateless at the moment, but it doesn't have to be, it can do a database lookup to figure out if `to` is an EVM actor, if necessary.

Reply:
> FWIW I was told it was best to avoid using the same key for both, but again unless they tell me why... 🤷‍♂️

At the time we thought reusing the same key to sign two different message structures weakened cryptographic security. But it might have been an unfounded fear as the underlying signature scheme is the same (ECDSA).

Reply:
> It is certainly possible to invoke an ethereum contract using a masked ID, which is an ethereum account hash that looks like 0xff0000...0123; such an address will go in the message.to as an f0 address. **There is some caveat but I forgot what it was.**

Two caveats: ID reorg sensitivity (in longest chain consensus chain, probably not a problem in immediate finality consensus), and address visible to contract (will be its natural Eth address, since the actor has a f410 delegated address in the state tree, and not the synthetic masked ID address).